### PR TITLE
Case Insensitive Support in Regexp Interval

### PR DESCRIPTION
**Pull request by mattweber**
### Description

Add a `case_insensitive` flag to regexp interval source.
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by mattweber** [comment]
cc @reta @andrross @dblock

File: server/src/test/java/org/opensearch/index/query/IntervalQueryBuilderTests.java [hunk 1]
```diff
@@ -845,8 +845,8 @@ public void testWildcard() throws IOException {
         });
     }
 
-    private static IntervalsSource buildRegexpSource(String pattern, int flags, Integer maxExpansions) {
```
Comment (nknize) [inline server/src/test/java/org/opensearch/index/query/IntervalQueryBuilderTests.java:848]: Let's keep this as an overloaded method that calls:

```
buildRegexpSource(pattern, flags, 0, maxExpansions)
```
I'm not a fan of the C style flags that lucene uses here because it's hard to follow (I always have to look at the lucene java docs :/ )

File: server/src/test/java/org/opensearch/index/query/IntervalQueryBuilderTests.java [hunk 2]
```diff
@@ -861,7 +861,7 @@ public void testRegexp() throws IOException {
         String json = "{ \"intervals\" : { \"" + TEXT_FIELD_NAME + "\": { " + "\"regexp\" : { \"pattern\" : \"te.m\" } } } }";
 
         IntervalQueryBuilder builder = (IntervalQueryBuilder) parseQuery(json);
-        Query expected = new IntervalQuery(TEXT_FIELD_NAME, buildRegexpSource("te.m", DEFAULT_FLAGS, null));
+        Query expected = new IntervalQuery(TEXT_FIELD_NAME, buildRegexpSource("te.m", DEFAULT_FLAGS, 0, null));
```
Comment (nknize) [inline server/src/test/java/org/opensearch/index/query/IntervalQueryBuilderTests.java:864]: By overloading `buildRegexpSource` as described above we should be able to keep this original call. At first look it's not clear what `0` does here and maintainers will have to constantly look back.

File: server/src/main/java/org/opensearch/index/query/IntervalsSourceProvider.java [hunk 3]
```diff
@@ -687,24 +688,35 @@ String getUseField() {
         private final int flags;
         private final String useField;
         private final Integer maxExpansions;
+        private final boolean caseInsensitive;
 
-        public Regexp(String pattern, int flags, String useField, Integer maxExpansions) {
+        public Regexp(String pattern, int flags, String useField, Integer maxExpansions, boolean caseInsensitive) {
```
Comment (nknize) [inline server/src/main/java/org/opensearch/index/query/IntervalsSourceProvider.java:693]: Can we add a javadoc here that references Lucene's (one and only; lol) [matching flag](https://github.com/apache/lucene/blob/d47ff38d703c6b5da1ef9c774ccda201fd682b8d/lucene/core/src/java/org/apache/lucene/util/automaton/RegExp.java#L413). 

FWIW, I hate the C style flag lucene uses here. Maybe this is a good first lucene commit for someone to simplify the matching flags of RegExp.java?

**Review by nknize — CHANGES_REQUESTED** [review]
looking good! just a couple comments

File: server/src/main/java/org/opensearch/index/query/IntervalsSourceProvider.java (hunk 3 above)
Comment (mattweber) [reply server/src/main/java/org/opensearch/index/query/IntervalsSourceProvider.java:699]: @nknize sure, I could work on a followup PR to lucene

**Review by nknize — APPROVED** [review]
nice! LGTM!

**Review by reta — APPROVED** [review]
